cpp/src/arrow/dataset/discovery.h

```diff
@@ -192,6 +192,10 @@ struct FileSystemFactoryOptions {
       ".",
       "_",
   };
+
+  /// when java context have a file system reference to be used
+  // then use this file system reference in the context
+  std::shared_ptr<void> file_system_java = nullptr;
```

Comment:
I don't think this needs to be Java-specific at all.

This should be `shared_ptr<FileSystem>`.